6/introduction.md

```diff
@@ -1,2 +1,10 @@
-PUBSUB
-======
\ No newline at end of file
+Pubsub - it's a party now
+=========================
+
+So far, our adventures in libp2p have all been about building nodes that only talk bidirectionally. In other words, only peer to peer and no group conversations.  We've built a secure chat protocol that works on multiple transport protocols and allows nodes to find each other in multiple ways, but it only allows you to hold a conversation with one other person.  
+
+Enter pubsub.  Pubsub is short for publish/subscribe and is a familiar concept in the world of distributed systems.  The basic premise is that there is a list of topics that nodes are interested in so they subscribe to these topics and can also publish messages under any given topic and that message will get relayed among all the nodes that are subscribed to that specific topic.  It's no longer just Alice and Bob talking face to face but Alice, Bob, Caroline, Dave, etc. all having a group SMS conversation about a specific topic.  
+
+The implementation of pubsub that libp2p uses is called gossipsub and the full implementation details can be found [here](https://docs.libp2p.io/concepts/publish-subscribe/).  The high-level idea of how this works is that as nodes join the network and discover one another using any of the various methods covered in section 4, they also query all other discovered nodes to see if they support pubsub, and if so, the new peers are added to the node's `peerStore` with the annotation that they support pubsub.  Once discovered, the node then creates a subset of these pubsub-supporting peers to be its full-message peers and it relays all published messages it receives to these peers.  As such, messages bounce around the network and peers retransmit messages they haven't seen before and this helps ensure that all nodes receive all messages.  
+
+So, let's get down to business.  In this section, we're going to add pubsub to our node's configuration, implement the details of it within the chat protocol, and roll out a full-on chat app, capable of support both direct and group messaging, all in a decentralized, secure environment not dependent on Facebook, WhatsApp, or any other centralized point of failure to get in the way.
```

Comment:
```suggestion
So far, our adventures in libp2p have all been about building nodes that only talk bidirectionally. In other words, only peer can talk to one peer and no group conversations.  
To solve this problem decentralised systems use Pubsub in short for publish/subscribe. A node can subscribe any interested topic to start receiving the messages. Once a node is subscribed to a topic it can also publish messages. Once a message is published under any given topic it gets relayed among all nodes subscribed to that topic.
 
Libp2p provides an implementation `pubsub` module called `gossipsub` and the full implementation details can be found [here](https://docs.libp2p.io/concepts/publish-subscribe/).  The high-level idea of how this works is that as nodes join the network and discover one another using any of the various methods covered in section 4, they also query all other discovered nodes to see if they support pubsub, and if so, the new peers are added to the node's `peerStore` with the annotation that they support pubsub.  Once discovered, the node then creates a subset of these pubsub-supporting peers to be its full-message peers and it relays all published messages it receives to these peers.  As such, messages bounce around the network and peers retransmit messages they haven't seen before and this helps ensure that all nodes receive all messages.  

So, let's get down to business.  In this section, we're going to add pubsub to our node's configuration, implement the details of it within the chat protocol, and roll out a full-on chat app, capable of support both direct and group messaging, all in a decentralised fashion unlike Facebook, WhatsApp, or any other centralised system which have a single point of failure.
```